File: warehouse/erp_the20/repositories/handover_repository.py
# -*- coding: utf-8 -*-
from __future__ import annotations
from typing import Optional, Dict, Any
from django.db import transaction
from django.db.models import QuerySet
from django.utils import timezone

from erp_the20.models import Handover, HandoverItem

# Mọi thao tác DB dùng alias này
ERP_DB_ALIAS = "erp_postgres"

# ========================= Helpers =========================
def _maybe_for_update(qs: QuerySet) -> QuerySet:
    """
    Chỉ thêm SELECT ... FOR UPDATE nếu connection(alias) đang ở trong atomic block.
    """
    try:
        conn = transaction.get_connection(using=ERP_DB_ALIAS)
        if getattr(conn, "in_atomic_block", False):
            return qs.select_for_update()
    except Exception:
        pass
    return qs

# ========================= Queries (đọc) =========================
def list_handover(filters: Optional[Dict[str, Any]] = None) -> QuerySet[Handover]:
    filters = filters or {}
    qs = Handover.objects.using(ERP_DB_ALIAS).all()
    if (v := filters.get("employee_id")) is not None:
        qs = qs.filter(employee_id=v)
    if (v := filters.get("manager_id")) is not None:
        qs = qs.filter(manager_id=v)
    if (v := filters.get("receiver_employee_id")) is not None:
        qs = qs.filter(receiver_employee_id=v)
    if (v := filters.get("status")) is not None:
        qs = qs.filter(status=v)
    return qs.order_by("-created_at").prefetch_related("items")

def get_handover(handover_id: int) -> Handover:
    return Handover.objects.using(ERP_DB_ALIAS).get(id=handover_id)

def get_item(item_id: int) -> HandoverItem:
    return HandoverItem.objects.using(ERP_DB_ALIAS).get(id=item_id)

# ========================= Mutations (ghi) =========================
def create_handover(**data) -> Handover:
    with transaction.atomic(using=ERP_DB_ALIAS):
        return Handover.objects.using(ERP_DB_ALIAS).create(**data)

def update_handover_status(handover_id: int, status: int) -> Handover:
    with transaction.atomic(using=ERP_DB_ALIAS):
        ho = _maybe_for_update(Handover.objects.using(ERP_DB_ALIAS)).get(id=handover_id)
        ho.status = status
        ho.save(using=ERP_DB_ALIAS, update_fields=["status", "updated_at"])
        return ho

def add_item(handover_id: int, *, title: str, detail: str = "", assignee_id: Optional[int] = None) -> HandoverItem:
    """
    FIX lõi: tạo item bằng handover_id (int) → gán FK bằng 'handover_id=handover_id'
    Tránh truyền object vào _id.
    """
    with transaction.atomic(using=ERP_DB_ALIAS):
        it = HandoverItem.objects.using(ERP_DB_ALIAS).create(
            handover_id=handover_id,
            title=title,
            detail=detail or "",
            assignee_id=assignee_id,
            # để repo không phụ thuộc enum, fallback về 0 nếu model không có enum
            status=getattr(HandoverItem.ItemStatus, "PENDING", 0),
        )
        # Đồng bộ trạng thái handover: nếu đang OPEN → IN_PROGRESS
        try:
            ho = get_handover(handover_id)
            if ho.status == getattr(Handover.Status, "OPEN", 0):
                ho.status = getattr(Handover.Status, "IN_PROGRESS", 1)
                ho.save(using=ERP_DB_ALIAS, update_fields=["status", "updated_at"])
        except Exception:
            pass
        return it

def set_item_status(item_id: int, status: int) -> HandoverItem:
    """
    Update trạng thái item; đồng bộ trạng thái Handover:
    - nếu tất cả DONE → Handover = DONE
    - nếu có item mà Handover còn OPEN → Handover = IN_PROGRESS
    """
    with transaction.atomic(using=ERP_DB_ALIAS):
        qs = HandoverItem.objects.using(ERP_DB_ALIAS).select_related("handover")
        qs = _maybe_for_update(qs)
        item = qs.get(id=item_id)

        item.status = status
        item.done_at = timezone.now() if status == getattr(HandoverItem.ItemStatus, "DONE", 1) else None
        item.save(using=ERP_DB_ALIAS, update_fields=["status", "done_at", "updated_at"])

        ho = item.handover
        total = ho.items.using(ERP_DB_ALIAS).count()
        done = ho.items.using(ERP_DB_ALIAS).filter(status=getattr(HandoverItem.ItemStatus, "DONE", 1)).count()

        if total and done == total:
            if ho.status != getattr(Handover.Status, "DONE", 2):
                ho.status = getattr(Handover.Status, "DONE", 2)
                ho.save(using=ERP_DB_ALIAS, update_fields=["status", "updated_at"])
        else:
            if ho.status == getattr(Handover.Status, "OPEN", 0):
                ho.status = getattr(Handover.Status, "IN_PROGRESS", 1)
                ho.save(using=ERP_DB_ALIAS, update_fields=["status", "updated_at"])

        return item

File: warehouse/erp_the20/repo/handover.repository.js
// repo/handover.repository.js
// list/get/create/update status, add item, set item status with handover status sync

module.exports = function createHandoverRepo({ Handover, HandoverItem, sequelize }) {
    function listHandover(filters = {}) {
      const where = {};
      if (filters.employee_id != null) where.employee_id = filters.employee_id;
      if (filters.manager_id != null) where.manager_id = filters.manager_id;
      if (filters.receiver_employee_id != null) where.receiver_employee_id = filters.receiver_employee_id;
      if (filters.status != null) where.status = filters.status;
  
      return Handover.findAll({
        where,
        order: [["created_at", "DESC"]],
        include: [{ model: HandoverItem, as: "items" }],
      });
    }
  
    async function getHandover(id) {
      const one = await Handover.findByPk(id, { include: [{ model: HandoverItem, as: "items" }] });
      if (!one) throw new Error("Handover not found");
      return one;
    }
  
    async function getItem(itemId) {
      const one = await HandoverItem.findByPk(itemId);
      if (!one) throw new Error("HandoverItem not found");
      return one;
    }
  
    async function createHandover(data) {
      return sequelize.transaction(async (t) => Handover.create(data, { transaction: t }));
    }
  
    async function updateHandoverStatus(handoverId, status) {
      return sequelize.transaction(async (t) => {
        const ho = await Handover.findByPk(handoverId, { transaction: t, lock: t.LOCK.UPDATE });
        if (!ho) throw new Error("Handover not found");
        ho.status = status;
        await ho.save({ transaction: t, fields: ["status", "updated_at"] });
        return ho;
      });
    }
  
    async function addItem(handoverId, { title, detail = "", assignee_id = null }) {
      return sequelize.transaction(async (t) => {
        const item = await HandoverItem.create(
          {
            handover_id: handoverId,
            title,
            detail: detail || "",
            assignee_id,
            status: HandoverItem.ItemStatus?.PENDING ?? 0,
          },
          { transaction: t }
        );
  
        try {
          const ho = await Handover.findByPk(handoverId, { transaction: t, lock: t.LOCK.UPDATE });
          if (ho && ho.status === (Handover.Status?.OPEN ?? 0)) {
            ho.status = Handover.Status?.IN_PROGRESS ?? 1;
            await ho.save({ transaction: t, fields: ["status", "updated_at"] });
          }
        } catch (_) {}
  
        return item;
      });
    }
  
    async function setItemStatus(itemId, status) {
      return sequelize.transaction(async (t) => {
        const item = await HandoverItem.findByPk(itemId, {
          include: [{ model: Handover, as: "handover" }],
          transaction: t,
          lock: t.LOCK.UPDATE,
        });
        if (!item) throw new Error("HandoverItem not found");
  
        item.status = status;
        item.done_at = status === (HandoverItem.ItemStatus?.DONE ?? 1) ? new Date() : null;
        await item.save({ transaction: t, fields: ["status", "done_at", "updated_at"] });
  
        const ho = item.handover;
        const total = await HandoverItem.count({ where: { handover_id: ho.id }, transaction: t });
        const done = await HandoverItem.count({ where: { handover_id: ho.id, status: (HandoverItem.ItemStatus?.DONE ?? 1) }, transaction: t });
  
        if (total && done === total) {
          if (ho.status !== (Handover.Status?.DONE ?? 2)) {
            ho.status = Handover.Status?.DONE ?? 2;
            await ho.save({ transaction: t, fields: ["status", "updated_at"] });
          }
        } else {
          if (ho.status === (Handover.Status?.OPEN ?? 0)) {
            ho.status = Handover.Status?.IN_PROGRESS ?? 1;
            await ho.save({ transaction: t, fields: ["status", "updated_at"] });
          }
        }
  
        return item;
      });
    }
  
    return { listHandover, getHandover, getItem, createHandover, updateHandoverStatus, addItem, setItemStatus };
  };